nestkernel/per_thread_bool_indicator.h

```diff
@@ -0,0 +1,105 @@
+/*
+ *  per_thread_bool_indicator.h
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef PER_THREAD_BOOL_INDICATOR_H
+#define PER_THREAD_BOOL_INDICATOR_H
+
+// C++ includes:
+#include <cassert>
+#include <cstddef>
+
+// Includes from nestkernel:
+#include "nest_types.h"
+#include "vp_manager.h"
+
+// Includes from sli:
+#include "dictdatum.h"
+
+namespace nest
+{
+
+class BoolIndicatorUInt64
+{
+public:
+  BoolIndicatorUInt64();
+  BoolIndicatorUInt64( const bool status );
+
+  bool is_true() const;
+  bool is_false() const;
+
+  void set_true();
+  void set_false();
+
+  void logical_and( const bool status );
+
+private:
+  static const uint_fast64_t true_uint64 = true;
+  static const uint_fast64_t false_uint64 = false;
+  uint_fast64_t status_;
+
+  friend class PerThreadBoolIndicator;
+};
+
+/**
+ * A thread-safe vector to keep track of the status across threads,
+ * for example during gather operations. Uses a vector of integers
+ * instead of a vector of bools to guarantee thread safety.
+ */
+class PerThreadBoolIndicator
+{
+public:
+  PerThreadBoolIndicator(){};
+
+  BoolIndicatorUInt64& operator[]( const thread tid );
```

Comment:
I would say: To be added as the need arises.

Reply:
if they'd be provided, wouldn't such calls as in https://github.com/nest/nest-simulator/blob/4b15e1aef8f86e74b8d60ef9fc8e0391764a9297/nestkernel/connection_manager.cpp#L558 use the const version? depends on what you mean by "need" ;)

Reply:
I think `check_primary_connections_` would have to be `const`  in order to result in a call to the additional constant `operator[]` function you are suggesting. But please feel free to try this out and, of course, adapt the code in case you are right. I just don't want to add a function that is currently not used.